File: src/Core/Math/VectorUtils.php
<?php
declare(strict_types=1);

namespace SMSolver\Core\Math;


class VectorUtils
{
    public static function roundVector(array &$vector, int $precision = 3): void
    {
        foreach ($vector as &$entry)
            $entry = round($entry, $precision);
    }

    public static function scalarMultiply(array &$vector, float $scalar): void
    {
        foreach ($vector as &$entry)
            $entry = $entry * $scalar;
    }

    /**
     * @param int $rows
     * @param int $columns
     * @return int[][]
     */
    public static function zerosMatrix(int $rows, int $columns): array
    {
        $M = array_fill(0, $rows, $columns);
        return array_map(fn(int $columns) => self::zerosVector($columns), $M);
    }

    /**
     * @param int $entries
     * @return int[]
     */
    public static function zerosVector(int $entries): array
    {
        return array_fill(0, $entries, 0);
    }
}

File: src/Core/Math/VectorUtils.go
package Math

import "math"

func RoundVector(vector []float64, precision int) []float64 {
	p := math.Pow10(precision)
	for i, e := range vector {
		vector[i] = math.Round(e*p) / p
	}
	return vector
}

func ScalarMultiply(vector []float64, scalar float64) []float64 {

	for i, e := range vector {
		vector[i] = e * scalar
	}
	return vector
}

func ZerosMatrix[T any](rows int, columns int) [][]T {
	m := make([][]T, rows)
	for i, _ := range m {
		m[i] = ZerosVector[T](columns)
	}
	return m
}
func ZerosVector[T any](entries int) []T {
	return make([]T, entries)
}
